src/java/org/apache/cassandra/locator/AbstractReplicaCollection.java

```diff
@@ -58,45 +63,332 @@
-    protected abstract C snapshot(List<Replica> subList);
-    protected abstract C self();
     /**
-     * construct a new Mutable of our own type, so that we can concatenate
-     * TODO: this isn't terribly pretty, but we need sometimes to select / merge two Endpoints of unknown type;
+     * A simple map that ensures the underlying list's iteration order is maintained, and can be shared with
+     * subLists (either produced via subList, or via filter that naturally produced a subList).
+     * This permits us to reduce the amount of garbage generated, by not unnecessarily copying,
+     * reduces the amount of indirection necessary, as well as ensuring monomorphic callsites.
+     * The underlying map is also more efficient, particularly for such small collections as we typically produce.
      */
-    public abstract Mutable<C> newMutable(int initialCapacity);
+    protected static class ReplicaMap<K> extends AbstractMap<K, Replica>
+    {
+        private final Function<Replica, K> toKey;
+        private final ReplicaList list;
+        // we maintain a map of key -> index in our list; this lets us share with subLists (or between Builder and snapshots)
+        // since we only need to corroborate that the list index we find is within the bounds of our list
+        // (if not, it's a shared map, and the key only occurs in one of our ancestors)
+        private final ObjectIntOpenHashMap<K> map;
+        private Set<K> keySet;
+        private Set<Entry<K, Replica>> entrySet;
+
+        abstract class AbstractImmutableSet<K> extends AbstractSet<K>
+        {
+            @Override
+            public boolean removeAll(Collection<?> c) { throw new UnsupportedOperationException(); }
+            @Override
+            public boolean remove(Object o) { throw new UnsupportedOperationException(); }
+            @Override
+            public int size() { return list.size(); }
+        }
+
+        class KeySet extends AbstractImmutableSet<K>
+        {
+            @Override
+            public boolean contains(Object o) { return containsKey(o); }
+            @Override
+            public Iterator<K> iterator() { return list.transformIterator(toKey); }
+        }
+
+        class EntrySet extends AbstractImmutableSet<Entry<K, Replica>>
+        {
+            @Override
+            public boolean contains(Object o)
+            {
+                Preconditions.checkNotNull(o);
+                if (!(o instanceof Entry<?, ?>)) return false;
+                return Objects.equals(get(((Entry) o).getKey()), ((Entry) o).getValue());
+            }
+
+            @Override
+            public Iterator<Entry<K, Replica>> iterator()
+            {
+                return list.transformIterator(r -> new SimpleImmutableEntry<>(toKey.apply(r), r));
+            }
+        }
+
+        private static <K> boolean putIfAbsent(ObjectIntOpenHashMap<K> map, Function<Replica, K> toKey, Replica replica, int index)
```

Comment:
not that it's terribly important, but we can apply `toKey` already and just get a key here.

Reply:
Thanks.  In this case I'm not convinced it would improve clarity, as we simply shift some complexity into the invocation.  However, I took another look and decided we could simply invoke `internalPutIfAbsent` in the constructor, so I've eliminated the static instance entirely.

Reply:
Sure, np